src/inmanta/ast/__init__.py

```diff
@@ -996,3 +995,50 @@ def format_trace(self, indent: str = "", indent_level: int = 0) -> str:
             out += "\n" + part
 
         return out
+
+
+class AttributeNotFound(NotFoundException, AttributeError):
+    """
+    Exception used for backwards compatibility with try-except blocks around some_proxy.some_attr.
+    This previously raised `NotFoundException` which is currently deprecated in this context.
+    Its new behavior is to raise an AttributeError for compatibility with Python's builtin `hasattr`.
+    """
+
+
+class UnsetException(RuntimeException):
+    """
+    This exception is thrown when an attribute is accessed that was not yet
+    available (i.e. it has not been frozen yet).
+    """
+
+    def __init__(self, msg: str, instance: Optional["Instance"] = None, attribute: Optional["Attribute"] = None) -> None:
+        RuntimeException.__init__(self, None, msg)
+        self.instance: Optional["Instance"] = instance
+        self.attribute: Optional["Attribute"] = attribute
+        self.msg = msg
+
+    def get_result_variable(self) -> Optional["Instance"]:
+        return self.instance
+
+
+class UnknownException(Exception):
+    """
+    This exception is thrown when code tries to access a value that is
+    unknown and cannot be determined during this evaluation. The code
+    receiving this exception is responsible for invalidating any results
+    depending on this value by return an instance of Unknown as well.
+    """
+
+    def __init__(self, unknown: "Unknown"):
+        super().__init__()
+        self.unknown = unknown
+
+
+@stable_api
+class PluginException(Exception):
```

Comment:
I don't like this location for the stable API though. Have to give this one some more thought.

Reply:
I agree to that